python/lsst/daf/butler/core/dimensions/sets.py

```diff
@@ -0,0 +1,558 @@
+        return NotImplemented
+
+    def __xor__(self, other):
+        if isinstance(other, DimensionSet):
+            return self.symmetric_difference(other)
+        return NotImplemented
+
+    def __sub__(self, other):
+        if isinstance(other, DimensionSet):
+            return self.difference(other)
+        return NotImplemented
+
+    def findIf(self, predicate, default=None):
+        """Return the element in ``self`` that matches the given predicate.
+
+        Parameters
+        ----------
+        predicate : callable
+            Callable that takes a single `DimensionElement` argument and
+            returns a `bool`, indicating whether the given value should be
+            returned.
+        default : `DimensionElement`, optional
+            Object to return if no matching elements are found.
+
+        Returns
+        -------
+        matching : `DimensionElement`
+            Element matching the given predicate.
+
+        Raises
+        ------
+        ValueError
+            Raised if multiple elements match the given predicate.
+        """
+        t = tuple(element for element in self if predicate(element))
+        if len(t) > 1:
+            raise ValueError(f"Multiple matches: {t}")
+        elif len(t) == 0:
+            return default
+        return t[0]
+
+
+class DimensionNameSet:
+    r"""An incomplete, name-only stand-in for `DimensionSet` or
+    `DimensionGraph`.
+
+    Because true `DimensionSet`\s and `DimensionGraph`\s cannot be constructed
+    without access to a "universe" `DimensionGraph` loaded from config,
+    requiring one of these classes in API also makes that API more difficult
+    to use.  `DimensionNameSet` partially solves that problem by being easy to
+    construct (only the names of the `DimensionElement`\s are needed, and no
+    sorting or checking is done) and behaving as much like a `DimensionSet` or
+    `DimensionGraph` as possible.  This enables the following pattern:
+
+     - Accept either `DimensionNameSet` as well as `DimensionSet` and/or
+       `DimensionGraph` when construting objects that need a container of
+       `DimensionElement`\s.  This may limit the functionality of the
+       constructed object if only a `DimensionNameSet` is passed, of course.
+
+     - "Upgrade" from `DimensionNameSet` to one of the more complete classes
+       when the object is rendezvouzed with a "universe" `DimensionGraph`.
+       This upgrade process also serves to validate the names.
+
+    The `DatasetType` class provides an example of this pattern;
+    `DatasetType`\s may be constructed with only the names of `Dimension`\s,
+    but are modified transparently by `Registry` operations to hold actual
+    `Dimension` objects.
+
+    Parameters
+    ----------
+    names : iterable of `str`
+        The names of elements to conceptually include in the set.
+    """
+
+    def __init__(self, names):
+        if not isinstance(names, Set):
+            names = frozenset(names)
```

Comment:
The conditional here is to avoid transforming a `OrderedDict` keys-view (which is a `Set`) into a `frozenset`, as that transformation discards the ordering.  But maybe I should also check that's not a `MutableSet` (which `OrderedDict` keys-view is not).  I'll do that and add some comments.